### PR TITLE
Add content on Dynamic Client Registration

### PR DESCRIPTION
**Pull request by mpaktiti**

### CODE REVIEW

File: articles/api-auth/dynamic-client-registration.md [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+### Update tenant settings
+
+In order to enable the feature, you need to enable the following flags at your tenant's settings:
+- `enable_dynamic_client_registration`: Enables the feature.
+- `enable_pipeline2`: Enables the underlying pipeline that is a required dependency.
+
+You can enable these flags using the [Update tenant settings endpoint](/api/management/v2#!/Tenants/patch_settings).
+
+```
+curl -X PATCH -H "Authorization: Bearer API2_ACCESS_TOKEN" -H "Content-Type: application/json" -H "Cache-Control: no-cache" -d '{
```
Comment (mpaktiti) [inline articles/api-auth/dynamic-client-registration.md]: Should I use `har` for this?

File: articles/api-auth/dynamic-client-registration.md [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+You need to update the `API2_ACCESS_TOKEN` with a valid Auth0 API2 token with the scope `update:tenant_settings`. See [How to get a Management APIv2 Token](/api/management/v2/tokens#how-to-get-a-management-apiv2-token) for details on how to do so.
+
+### Promote connections
+
+Clients registered via the [Dynamic Client Registration Endpoint](#dynamic-client-registration-endpoint) are flagged as **Third Party Clients** and can only authenticate users using connections flagged as **Domain Connections**. These connections will be open for any dynamic client to allow users to authenticate.
+
+You can promote a connection to domain level using the [Update a Connection endpoint](/api/management/v2#!/Connections/patch_connections_by_id).
+
+```
```
Comment (mpaktiti) [inline articles/api-auth/dynamic-client-registration.md]: Should I use `har` for this?

File: articles/api-auth/dynamic-client-registration.md [hunk 3]
```diff
@@ -0,0 +1,205 @@
+---
+description: How to dynamically register Clients with Auth0, using the Management APIv2.
+toc: true
+---
+
+# Dynamic Client Registration
+
+Dynamic Client Registration is a feature, based on the [OpenID Connect Dynamic Client Registration specification](https://openid.net/specs/openid-connect-registration-1_0.html), that enables you to to register clients dynamically, using the [Management APIv2](/api/management/v2). This way you can allow third party developers to register and use clients.
```
Comment (cabralmartin) [inline articles/api-auth/dynamic-client-registration.md]: s/to to register/to register

File: articles/api-auth/dynamic-client-registration.md [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+All clients, registered dynamically with Auth0, have the following characteristics:
+
+- The [ID tokens](/tokens/id-token) generated for these clients, hold minimum user profile information. In order to retrieve the full user profile, you need to get an access token and use the [/userinfo endpoint](/api/authentication#get-user-info).
+
+- These clients cannot use [ID tokens](/tokens/id-token) to invoke [Management APIv2](/api/management/v2) endpoints. Instead, they should [get a Management APIv2 Token](/api/management/v2/tokens#how-to-get-a-management-apiv2-token). Note that the client should be granted the `current_user_*` scopes, as required by each endpoint.
```
Comment (cabralmartin) [inline articles/api-auth/dynamic-client-registration.md]: The linked document shows how to obtained such scope thru a client credentials exchange. That exchange won't work because that's authorization for clients; the notion of user is not present there.

To get these scopes, clients will need to make a call to `/authorize` with the following,
- `audience=API2_URL`
- `scopes=read:current_user update:current_user_metadata`

This will give the client an access token to API2 with the selected scopes for that particular user.

File: articles/api-auth/dynamic-client-registration.md [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+  - `update:current_user_metadata`: [Update a user](/api/management/v2#!/Users/patch_users_by_id), [Delete a user's multifactor provider](/api/management/v2#!/Users/delete_multifactor_by_provider)
+  - `delete:current_user_metadata`: [Delete a user](/api/management/v2#!/Users/delete_users_by_id)
+  - `create:current_user_metadata`: [Create a user](/api/management/v2#!/Users/post_users)
+  - `create:current_user_device_credentials`: [Create a device public key](/api/management/v2#!/Device_Credentials/post_device_credentials)
+  - `delete:current_user_device_credentials`: [Delete a device credential](/api/management/v2#!/Device_Credentials/delete_device_credentials_by_id)
+  - `update:current_user_identities`: [Link a user account](/api/management/v2#!/Users/post_identities), [Unlink a user identity](/api/management/v2#!/Users/delete_provider_by_user_id)
+
+- They cannot have any connections, except for tenant level connections (domain connections). To use tenant level connections, you need the latest version of [Lock](/libraries/lock).
```
Comment (cabralmartin) [inline articles/api-auth/dynamic-client-registration.md]: Rephrase. `If the client wants to use Lock to authenticate users, it will need to use a version greater than 10.7`

Note, clients don't _have_ connections. They can consume a connection, but it's not like an intrinsic property of the client. The picture gets clearer with "Domain Connections". These are consumed by third party clients but they have no other relationship whatsoever.

File: articles/api-auth/dynamic-client-registration.md (hunk 2 above)
Comment (cabralmartin) [reply articles/api-auth/dynamic-client-registration.md]: Yes, if we can change all the curls to HARs that will be a huge win.

File: articles/api-auth/dynamic-client-registration.md [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+</script>
+```
+
+## Use dynamic registration
+
+In this section we will see how a third party developer can register and configure a client.
+
+### Register your client
+
+In order to dynamically register a client with Auth0, you need to send an HTTP `POST` message to the [Client Registration endpoint](/api/management/v2#!/oidc/register). Note that Auth0 supports open Dynamic Registration, which means that the [Client Registration endpoint](/api/management/v2#!/oidc/register) will accept a registration request without an [OAuth 2.0 Access Tokens](/tokens/access-token).
```
Comment (cabralmartin) [inline articles/api-auth/dynamic-client-registration.md]: This link does not exist `/api/management/v2#!/oidc/register`. We don't have swagger for this endpoint.

File: articles/api-auth/dynamic-client-registration.md [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+  - `create:current_user_metadata`: [Create a user](/api/management/v2#!/Users/post_users)
+  - `create:current_user_device_credentials`: [Create a device public key](/api/management/v2#!/Device_Credentials/post_device_credentials)
+  - `delete:current_user_device_credentials`: [Delete a device credential](/api/management/v2#!/Device_Credentials/delete_device_credentials_by_id)
+  - `update:current_user_identities`: [Link a user account](/api/management/v2#!/Users/post_identities), [Unlink a user identity](/api/management/v2#!/Users/delete_provider_by_user_id)
+
+- They cannot have any connections, except for tenant level connections (domain connections). To use tenant level connections, you need the latest version of [Lock](/libraries/lock).
+
+## Enable dynamic registration
+
+In this section we will see how you can enable the dynamic registration feature for your tenant.
```
Comment (cabralmartin) [inline articles/api-auth/dynamic-client-registration.md]: Let's make more explicit that if you turn on this feature, *anyone* can create clients in your tenant.

It's important to stress that before moving on to turning on the feature.

File: articles/api-auth/dynamic-client-registration.md [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+}
+```
+
+Where:
+- `client_name`: The name of the Dynamic Client to be created.
+- `redirect_uris`: An array of URLs that Auth0 will deem valid to call at the end of an Authentication flow.
+
+Optionally, you can set a value for `token_endpoint_auth_method`, which can be `none` or `client_secret_post` (default value).
+
+**NOTE**: You can set several attributes. For a complete listing refer to the [/oidc/register endpoint](/api/management/v2#!/oidc/register).
```
Comment (cabralmartin) [inline articles/api-auth/dynamic-client-registration.md]: Remove this. Also, the only attributes we allow in the payload are,
- client_name (required)
- redirect_urs (required)
- token_endpoint_auth_method (optional) (default: client_secret_post)

File: articles/api-auth/dynamic-client-registration.md (hunk 3 above)
Comment (cabralmartin) [inline articles/api-auth/dynamic-client-registration.md]: Note that while the code lives in the same deployment, we don't consider the dynamic registration endpoint part of API2. API2 endpoints live under `/api/v2/something` while the new endpoint is just `tenant.auth0.com/oidc/register`.

File: articles/clients/index.md [hunk 9]
```diff
@@ -83,3 +83,7 @@ Once you click on the button a pop-up window will ask you to confirm the action.
 Auth0 stores log data of both actions taken in the dashboard by the administrators, as well as authentications made by your users. The logs include many of the actions performed by the user like failing to login to a client or requesting a password change. For more details refer to: [Logs](/logs).
 
 If you use a third-party application for log management, like Sumo Logic, Splunk or Loggly, you can use Auth0 Extensions to export your logs there. For details on the available extensions and how to configure them refer to: [Extensions](/extensions).
+
+## Dynamic Client Registration
+
+You can use the [Management APIv2](/api/management/v2) to programmatically create clients, as described in the [OIDC Dynamic Client Registration 1.0 specification](https://openid.net/specs/openid-connect-registration-1_0.html). For more details please refer to [Dynamic Client Registration](/api-auth/dynamic-client-registration).
```
Comment (cabralmartin) [inline articles/clients/index.md]: s/You can use the api/You can use Auth0

(dynamic client is not part of API2)

**Review by cabralmartin — CHANGES_REQUESTED** [review]

File: articles/api-auth/dynamic-client-registration.md [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+toc: true
+---
+
+# Dynamic Client Registration
+
+Dynamic Client Registration is a feature, based on the [OpenID Connect Dynamic Client Registration specification](https://openid.net/specs/openid-connect-registration-1_0.html), that enables you to register clients dynamically. This way you can allow third party developers to register and use clients.
+
+All clients, registered dynamically with Auth0, have the following characteristics:
+
+- The [ID tokens](/tokens/id-token) generated for these clients, hold minimum user profile information. In order to retrieve the full user profile, you need to get an access token and use the [/userinfo endpoint](/api/authentication#get-user-info).
```
Comment (cabralmartin) [inline articles/api-auth/dynamic-client-registration.md]: /userinfo and id_token will return the same information (OIDC valid claims + (for idtoken) custom namespaced claims).

File: articles/api-auth/dynamic-client-registration.md [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+  - `update:current_user_metadata`: [Update a user](/api/management/v2#!/Users/patch_users_by_id), [Delete a user's multifactor provider](/api/management/v2#!/Users/delete_multifactor_by_provider)
+  - `delete:current_user_metadata`: [Delete a user](/api/management/v2#!/Users/delete_users_by_id)
+  - `create:current_user_metadata`: [Create a user](/api/management/v2#!/Users/post_users)
+  - `create:current_user_device_credentials`: [Create a device public key](/api/management/v2#!/Device_Credentials/post_device_credentials)
+  - `delete:current_user_device_credentials`: [Delete a device credential](/api/management/v2#!/Device_Credentials/delete_device_credentials_by_id)
+  - `update:current_user_identities`: [Link a user account](/api/management/v2#!/Users/post_identities), [Unlink a user identity](/api/management/v2#!/Users/delete_provider_by_user_id)
+
+::: panel-info How to get a Management APIv2 Token
+In order to access the [Management APIv2](/api/management/v2) endpoints, from a dynamic client, you need a Management APIv2 Token. To do so, invoke `https://${account.namespace}/authorize`, with the following request parameters:
+- `audience=API2_URL`
```
Comment (cabralmartin) [inline articles/api-auth/dynamic-client-registration.md]: I think we can populate the API2_URL maybe,
`https://${account.namespace}/api/v2/` (the last trailing slash needs to be there)

File: articles/api-auth/dynamic-client-registration.md [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+  - `update:current_user_metadata`: [Update a user](/api/management/v2#!/Users/patch_users_by_id), [Delete a user's multifactor provider](/api/management/v2#!/Users/delete_multifactor_by_provider)
+  - `delete:current_user_metadata`: [Delete a user](/api/management/v2#!/Users/delete_users_by_id)
+  - `create:current_user_metadata`: [Create a user](/api/management/v2#!/Users/post_users)
+  - `create:current_user_device_credentials`: [Create a device public key](/api/management/v2#!/Device_Credentials/post_device_credentials)
+  - `delete:current_user_device_credentials`: [Delete a device credential](/api/management/v2#!/Device_Credentials/delete_device_credentials_by_id)
+  - `update:current_user_identities`: [Link a user account](/api/management/v2#!/Users/post_identities), [Unlink a user identity](/api/management/v2#!/Users/delete_provider_by_user_id)
+
+::: panel-info How to get a Management APIv2 Token
+In order to access the [Management APIv2](/api/management/v2) endpoints, from a dynamic client, you need a Management APIv2 Token. To do so, invoke `https://${account.namespace}/authorize`, with the following request parameters:
```
Comment (cabralmartin) [inline articles/api-auth/dynamic-client-registration.md]: We can say,

In order to ... you can use any of the API Authorization Flows described here (link to docs) using the following params,
- audience: ....

File: articles/api-auth/dynamic-client-registration.md [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+
+# Dynamic Client Registration
+
+Dynamic Client Registration is a feature, based on the [OpenID Connect Dynamic Client Registration specification](https://openid.net/specs/openid-connect-registration-1_0.html), that enables you to register clients dynamically. This way you can allow third party developers to register and use clients.
+
+All clients, registered dynamically with Auth0, have the following characteristics:
+
+- The [ID tokens](/tokens/id-token) generated for these clients, hold minimum user profile information. In order to retrieve the full user profile, you need to get an access token and use the [/userinfo endpoint](/api/authentication#get-user-info).
+
+- They cannot use any connections, except for tenant level connections (domain connections). If the client wants to use [Lock](/libraries/lock) to authenticate users, it will need to use a version greater than `10.7`.
```
Comment (cabralmartin) [inline articles/api-auth/dynamic-client-registration.md]: Maybe we can elaborate on what a tenant connection is (its pretty much the same but is enabled for all 3rd party clients and not turned on for "specific" clients, etc).

File: articles/api-auth/dynamic-client-registration.md [hunk 14]
```diff
@@ -0,0 +1,232 @@
+---
+description: How to dynamically register Clients with Auth0, using the Management APIv2.
+toc: true
+---
+
+# Dynamic Client Registration
+
+Dynamic Client Registration is a feature, based on the [OpenID Connect Dynamic Client Registration specification](https://openid.net/specs/openid-connect-registration-1_0.html), that enables you to register clients dynamically. This way you can allow third party developers to register and use clients.
+
+All clients, registered dynamically with Auth0, have the following characteristics:
```
Comment (cabralmartin) [inline articles/api-auth/dynamic-client-registration.md]: Can we start using `Third Party Clients` to identified these? (instead of "registered dynamically"). This will help to associate concepts.

Also, wdyt about converting this to a sub section that describes 3rd party clients in detail? Maybe later it can be moved to its own doc.

File: articles/api-auth/dynamic-client-registration.md [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+  - `read:current_user`: [List or search users](/api/management/v2#!/Users/get_users), [Get a user](/api/management/v2#!/Users/get_users_by_id), [Get user Guardian enrollments](/api/management/v2#!/Users/get_enrollments)
+  - `update:current_user_metadata`: [Update a user](/api/management/v2#!/Users/patch_users_by_id), [Delete a user's multifactor provider](/api/management/v2#!/Users/delete_multifactor_by_provider)
+  - `delete:current_user_metadata`: [Delete a user](/api/management/v2#!/Users/delete_users_by_id)
+  - `create:current_user_metadata`: [Create a user](/api/management/v2#!/Users/post_users)
+  - `create:current_user_device_credentials`: [Create a device public key](/api/management/v2#!/Device_Credentials/post_device_credentials)
+  - `delete:current_user_device_credentials`: [Delete a device credential](/api/management/v2#!/Device_Credentials/delete_device_credentials_by_id)
+  - `update:current_user_identities`: [Link a user account](/api/management/v2#!/Users/post_identities), [Unlink a user identity](/api/management/v2#!/Users/delete_provider_by_user_id)
+
```
Comment (cabralmartin) [inline articles/api-auth/dynamic-client-registration.md]: Another key feature is that 3rd party clients CANNOT skip user consent when consuming APIs. That's the security of the feature; anyone can create clients but each client relays on the final user to accept consent. That's why this works.

File: articles/api-auth/index.html [hunk 16]
```diff
@@ -129,6 +129,12 @@
     </ul>
   </li>
   <li>
+    <i class="icon icon-budicon-715"></i><a href="/api-auth/dynamic-client-registration">Dynamic Client Registration</a>
+    <p>
+      Learn how to dynamically register Clients with Auth0, using the Management APIv2.
```
Comment (cabralmartin) [inline articles/api-auth/index.html]: Learn how to allow third party developers to create clients under your account following the OIDC Dynamic Client Registration specification.

File: articles/api-auth/dynamic-client-registration.md (hunk 14 above)
Comment (mpaktiti) [reply articles/api-auth/dynamic-client-registration.md]: Yes, that's better 👍

File: articles/api-auth/dynamic-client-registration.md [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+Dynamic Client Registration is a feature, based on the [OpenID Connect Dynamic Client Registration specification](https://openid.net/specs/openid-connect-registration-1_0.html), that enables you to register clients dynamically. These clients are called **Third Party Clients** (since they are registered and used by third party developers).
+
+## Third Party Client characteristics
+
+Third Party Clients have the following characteristics:
+
+- The [ID tokens](/tokens/id-token) generated for these clients, hold minimum user profile information.
+
+- They can use only tenant level connections (domain connections). These are sources of users, configured in the tenant's [dashboard](${manage_url}) as connections, but they are not enabled for any client.
```
Comment (cabralmartin) [inline articles/api-auth/dynamic-client-registration.md]: > they are not enabled for any client.

I think we are mentioning that they don't need to be explicitly enable for a particular client. When I read this tho, it sounds like this type of connection cannot be used by any of the clients.

Maybe rephrase? Make it clear that Domain Connection is enable for every third party client and that optionally it can be enabled for selected first party (standard) clients.

File: articles/api-auth/dynamic-client-registration.md [hunk 18]
```diff
@@ -0,0 +1,236 @@
...
+
+::: panel-warning Security warning
+Auth0 supports Open Dynamic Registration, which means that if you enable this feature, **anyone** will be able to create clients in your tenant without a token.
+:::
+
+### Update tenant settings
+
+In order to enable the feature, you need to set to `true` the following flags at your tenant's settings:
+- `enable_dynamic_client_registration`: Enables the feature.
+- `enable_pipeline2`: Enables the underlying pipeline that is a required dependency.
```
Comment (cabralmartin) [inline articles/api-auth/dynamic-client-registration.md:49]: Note to future self: This flag (`enable_pipeline2`) will be removed in the future. Let's keep on the radar that this doc will need to be updated once we finally merge the pipeline.

Comment (mpaktiti) [reply articles/api-auth/dynamic-client-registration.md:49]: I added a reminder for March.

File: articles/api-auth/dynamic-client-registration.md [hunk 19]
```diff
@@ -0,0 +1,236 @@
...
+
+- To authenticate users using [Lock](/libraries/lock), you will have to use a version greater than `10.7`.
+
+- They cannot skip user consent when consuming APIs. This is for security purposes, as anyone can create a client, but each client relays on the final user to provide consent.
+
+- They cannot use [ID tokens](/tokens/id-token) to invoke [Management APIv2](/api/management/v2) endpoints. Instead, they should get a Management APIv2 Token (see the *How to get a Management APIv2 Token* panel for details). Note that the client should be granted the `current_user_*` scopes, as required by each endpoint.
+  - `read:current_user`: [List or search users](/api/management/v2#!/Users/get_users), [Get a user](/api/management/v2#!/Users/get_users_by_id), [Get user Guardian enrollments](/api/management/v2#!/Users/get_enrollments)
+  - `update:current_user_metadata`: [Update a user](/api/management/v2#!/Users/patch_users_by_id), [Delete a user's multifactor provider](/api/management/v2#!/Users/delete_multifactor_by_provider)
+  - `delete:current_user_metadata`: [Delete a user](/api/management/v2#!/Users/delete_users_by_id)
```
Comment (cabralmartin) [inline articles/api-auth/dynamic-client-registration.md]: Remove `delete:current_user_metadata` and `create:current_user_metadata` from the list. While auth0-server will return these, they don't have a proper use case in API2.

**Review by cabralmartin — CHANGES_REQUESTED** [review]
